### PR TITLE
Add "dry run" option for testing deployments 

### PR DESCRIPTION
**Pull request by duncanhall**
To allow potential deployments to be tested without side-effects, this change adds an optional `dry_run` query parameter to `POST /api/vi/deployments` calls. 

Passing a value of `dry_run=true` will validate the provided deployment configuration (and any implicit configuration) but return before any changes are made. An option has been added to include this in the deployment modal UI:

selection_026

This change also includes fairly extensive test additions for `DeployService` - we previously had zero test coverage for our deployment command!

### CODE REVIEW

File: server/commands/deployments/DeployService.js [hunk 1]
```diff
@@ -44,11 +54,51 @@ module.exports = function DeployServiceCommandHandler(command) {
 
 function validateCommandAndCreateDeployment(command) {
   return co(function* () {
+    const { mode, packagePath, environmentName, serviceSlice, serviceName, serviceVersion } = command;
```
Comment (filso) [inline server/commands/deployments/DeployService.js]: that's nice. You could also put that as a function argument if you prefer, and remove "command" var

File: server/commands/deployments/DeployService.js [hunk 2]
```diff
@@ -44,11 +54,51 @@ module.exports = function DeployServiceCommandHandler(command) {
...
+      throw new Error('Slice must be set to \'none\' in overwrite mode.');
+    }
+    if (SupportedDeploymentModes.indexOf(mode.toLowerCase()) < 0) {
+      throw new Error(`Unknown mode \'${mode}\'. Supported modes are: ${SupportedDeploymentModes.join(', ')}`);
+    }
+    if (mode === 'bg' && SupportedSliceNames.indexOf(serviceSlice) < 0) {
+      throw new Error(`Unknown slice \'${serviceSlice}\'. Supported slices are: ${SupportedSliceNames.join(', ')}`);
+    }
+
+    if (!packagePath) {
```
Comment (filso) [inline server/commands/deployments/DeployService.js]: I think it's better to use strong comparisons

File: server/commands/deployments/DeployService.js [hunk 3]
```diff
@@ -44,11 +54,51 @@ module.exports = function DeployServiceCommandHandler(command) {
...
+        s3Package = yield s3PackageLocator.findDownloadUrl({
+          environment: environmentName,
+          service: serviceName,
+          version: serviceVersion
+        });
+      } catch (error) {
+        throw new Error(`An attempt to locate the following package in S3 was forbidden: ${serviceName} version ${serviceVersion}`);
+      }
+
+      if (!s3Package) {
```
Comment (filso) [inline server/commands/deployments/DeployService.js]: I think it's better to use strong comparisons

File: server/test/commandHandlers/deployment/DeployServiceTest.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
+/* Copyright (c) Trainline Limited, 2016-2017. All rights reserved. See LICENSE.txt in the project root for license information. */
+
+'use strict';
+
+const assert = require('assert')
+const rewire = require('rewire');
+const sinon = require('sinon');
+const _ = require('lodash');
+
+describe.only('DeployService', function() {
```
Comment (filso) [inline server/test/commandHandlers/deployment/DeployServiceTest.js]: Need to remove .only flag?
